core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -3242,6 +3209,133 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  def handleEnvelope(request: RequestChannel.Request): Unit = {
+    val envelope = request.body[EnvelopeRequest]
+
+    if (!config.metadataQuorumEnabled) {
+      // If forwarding is not yet enabled, we treat the request as unparsable and close the connection
+      closeConnection(request, Collections.emptyMap())
+      return
+    } else if (!request.context.fromPrivilegedListener) {
+      sendErrorResponseMaybeThrottle(request, new ClusterAuthorizationException(
```

Comment:
not sure whether ```ClusterAuthorizationException ``` suits this case. How about adding a (new) specify exception for this case?

Reply:
Makes sense. Initially I was doing a match on the result and returning directly, but it was verbose and awkward.

Reply:
To be honest, I'm more inclined to get rid of this requirement. It seems a little arbitrary that we require the inter-broker listener for envelope requests only. If a principal has CLUSTER_ACTION, I think the damage they can do already is virtually unlimited. 

@abbccdda What do you think? Do we have a strong reason for this?

Reply:
Perhaps I can restore the original behavior and just close the connection in order to unblock the PR. Then we can consider this question separately.

Reply:
> Perhaps I can restore the original behavior and just close the connection in order to unblock the PR. Then we can consider this question separately.

+1

Reply:
IIRC, the intend was to distinguish forwarded requests from non-forwarded requests in a prior version of the KIP when we were not using an envelope. I don't think that this is really necessary since we use an envelope now. At least, I can't think of a strong case for it.